Cover failed assertions

hannes-steffenhagen-diffblue opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

This adds an option `--cover-failed-assertions` that prevents the
default behaviour of coverage stopping at failed assertions by turning
assertions into skips rather than assumes (which is the default
behaviour for coverage criteria other than `assertion`, which behaves
the same with or without the flag).


This is one way to address https://github.com/diffblue/cbmc/issues/5543

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue committed:
Extract constants for --cover help message and options

These were previously duplicated between 3 different executables.

hannes-steffenhagen-diffblue committed:
Add --cover-failed-assertions option

This adds an option `--cover-failed-assertions` that prevents the
default behaviour of coverage stopping at failed assertions by turning
assertions into skips rather than assumes (which is the default
behaviour for coverage criteria other than `assertion`, which behaves
the same with or without the flag).

hannes-steffenhagen-diffblue commented on src/goto-instrument/cover.cpp:
```diff
@@ -285,16 +291,22 @@ static void instrument_cover_goals(
+        }
+        else
+        {
+          i_it->turn_into_skip();
```
Btw we have to skip them rather than just leaving them be because otherwise cbmc would consider these assertions coverage goals.

peterschrammel replied:
Why not make `--no-assertions` (aka turn-assertions-into-skips) useable with `--cover` then (as @martin-cs  suggests)? That would make the behaviour more transparent to the user.

hannes-steffenhagen-diffblue committed:
Update cbmc README.md with --cover-failed-assertions flag

hannes-steffenhagen-diffblue commented on src/cbmc/README.md:
```diff
@@ -76,13 +76,15 @@ by other mechanism may categorise their properties differently.
-proceeds the same as in all-properties mode. Coverage solving is implemented by
-`bmc_covert`, but is structurally practically identical to
+equation solver; In cases where this behaviour is undesirable you can pass the
+`--cover-failed-assertions` which makes coverage checking continue even for
```
BTW shouldn’t this file be under `doc` rather than `src`? I don’t really understand our documentation structure...

peterschrammel replied:
The READMEs which are in each module are pulled into here: http://cprover.diffblue.com/folder-walkthrough.html (click on `cbmc`, for example).
We decided to keep them in the source folders so that they are as close as possible to the code.

martin-cs commented on src/goto-instrument/cover.h:
```diff
@@ -22,9 +22,17 @@ class message_handlert;
-  " --cover CC                   create test-suite with coverage criterion CC\n"
+#define OPT_COVER                                                              \
+  "(cover):"                                                                   \
+  "(cover-failed-assertions)"
```
Is it possible to have this as an argument to `--cover`?

thomasspriggs replied:
:pick: If we keep this as a separate option from `--cover` then we should check that `--cover-failed-assertions` is not passed without `--cover` and exit with a suitable error message in the case of that particular invalid usage.

martin-cs commented on src/goto-instrument/cover.h:
```diff
@@ -22,6 +22,10 @@ class message_handlert;
 class cmdlinet;
 class optionst;
 
+#define OPT_COVER "(cover):"
```
Thank you for this refactoring.  Finding an removing unnecessary and redundant repetition is always good.

martin-cs commented on src/cbmc/README.md:
```diff
@@ -76,13 +76,15 @@ by other mechanism may categorise their properties differently.
-There is one further BMC mode that differs more fundamentally: when `--cover` is
-passed, assertions in the program text are converted into assumptions (these
+There is one further BMC mode that differs more fundamentally: when `--cover`
+is passed, assertions in the program text are converted into assumptions (these
```
I don't want to open a massive can of worms here but ... why is this the default?  Is this actually useful?  Given that there are options for `--no-assertions`, `--no-assumptions` and `--assert-to-assume` (whether they work or even do anything at all ... is an open question) could we make the transformation of assertions and assumptions orthogonal to `--cover`.  It feels like this should be possible and even desirable...

hannes-steffenhagen-diffblue replied:
@martin-cs I think this is the default because whoever wrote this had the expectation that programs should stop at failed assertions. I suppose we could change it, but I’m always a bit wary about changing defaults.

hannes-steffenhagen-diffblue replied:
@martin-cs FWIW because cover treats each failed goal as something relating to coverage right now we have to do _something_ with asserts. That said, I think the alternative (just making cover cleverer about how it counts coverage) would be more desirable.

I'd consider that as something we can take a look at in terms of future work though.

martin-cs replied:
Yes, if you are doing a naive count then, I see the need to do something with pre-existing assertions or otherwise true assertions (generally a good thing) will count against your coverage.  Smarter counting would certainly be one way to address this.

It feels to me that making the "doing something about asserts" independent might be a desirable option as well.  I can see use-cases for converting to SKIP, to ASSUME and leaving as they are.

Then again, I am not a user of this code at the moment and do not have time to work on it so my views maybe not so important.

martin-cs reviewed (approved):
Approve because the refactor is good.  I would prefer if the new option could be added in to the argument for `--cover` and I would prefer it even more if we could make assertion transformation orthogonal to and preceeding coverage assertions or at least have a conversation about it but I also realise that you need to get stuff done so approve.

NlightNFotis committed:
Update regexes in test to match actual lines matched.

NlightNFotis committed:
Mark new tests as expected failures for paths flag.

The tests are failing when the `--paths` flag is set.
This is because we understand that the paths code hasn't
been exercised thoroughly, and as a result this behaviour
may or may not indicate a potential issue -  we need to
investigate this at a later point.

hannes-steffenhagen-diffblue committed:
Do not check XML traces on --cover tests

peterschrammel commented on regression/validate-trace-xml-schema/check.py:
```diff
@@ -34,6 +34,9 @@
     ['unknown-argument-suggestion', 'test.desc'],
     # this one produces XML intermingled with main XML output when used with --xml-ui
     ['graphml_witness2', 'test.desc'],
+    # these are producing coverage goals which aren't including in the schema
```
including -> included?

thomasspriggs commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -73,8 +75,9 @@ class optionst;
   "(nondet-static)" \
-  "(version)" \
-  "(cover):(symex-coverage-report):" \
+  "(version)"        \
```
:grey_question: Is this additional spacing accidental?

hannes-steffenhagen-diffblue commented:
@peterschrammel `--no-assertions` applies to `goto_check` assertions, it only turns user provided assertions into skips. What we'd want for cover is to turn _all_ non-cover related assertions into skips, user added or not. Making it work differently depending on whether or not `--cover` is passed would be confusing.

That said, maybe the better solution would be to make sure cover only checks cover-added assertions for coverage metrics instead of this skip-workaround...

peterschrammel commented:
...unless you use `--cover assertion`.

hannes-steffenhagen-diffblue commented:
I don’t think the behaviour is any different for `--cover assertion`? It still only removes user assertions (from what I can tell `--cover assertions` doesn’t care specifically about this option at all, this is all happening in goto check), and the original complaint this is meant to address is that coverage changes based on whether or not you pass in the various `*-check` flags (which I think may well be expected behaviour in some circumstances, but is clearly not the only reasonable behaviour).

It’s just that (only) removing user assertions is actually what we want for `--cover assertion` I guess.

thomasspriggs commented on regression/cbmc/cover-failed-assertions/test-no-failed-assertions.desc:
```diff
@@ -0,0 +1,11 @@
+^EXIT=0$
+^SIGNAL=0$
+--
+^warning: ignoring
```
:pick: It would be nice to have a comment in each of `.desc` files explaining what the test is intended to test. Even something pretty short could make this a while lot easier to maintain.

thomasspriggs commented on src/goto-instrument/cover.h:
```diff
@@ -22,9 +22,17 @@ class message_handlert;
+#define HELP_COVER                                                             \
+  " --cover CC                   create test-suite with coverage criterion "   \
+  "CC\n"                                                                       \
+  " --cover-failed-assertions    do not stop coverage checking at failed "     \
```
:pick: Because of the preceding refactor, this option now applies to `jdiff` and `goto-diff` as well as `cbmc`. This means that we should probably test that it works with those entry points.

thomasspriggs reviewed (approved):
Seems reasonable. If it was up to me, I would squash the commits fixing the tests back into the commit where they were added. This is because it could make use of `git bisect` easier, if we ever need to use it. Please fix the spelling error noted by Peter before this is merged.

hannes-steffenhagen-diffblue committed:
Fix typo

hannes-steffenhagen-diffblue committed:
Fix spacing

hannes-steffenhagen-diffblue committed:
Add comments to tests

peterschrammel reviewed (approved):
I understand that this needs a quick fix. But I don't think that the current solution will be the last. The options around goto check assertions, user assertions, cover, the various -no-X options, etc need to be made consistent so that they can be used in a clearly understandable and orthogonal way without subtle side effects.

martin-cs commented:
I realise this is easy for me to say as 1. it's not me doing the work and 2. it's not me in need of the urgent fix but... +1 agree with @peterschrammel .  This has been a mess for too long.  I will leave it up to others (those that 1 and 2 apply to) to decide how this PR should relate to that work.
